browser/net/brave_oauth_protocol_handler.cc

```diff
@@ -0,0 +1,225 @@
+#include "brave/components/gemini/browser/gemini_service.h"
+#endif
+
+namespace {
+
+const std::map<std::string, std::string> redirects = {
+  {kBinanceScheme, "chrome://newtab?binanceAuth=1"},
+  {kGeminiScheme, "chrome://newtab?geminiAuth=1"},
+  {kRewardsScheme, "chrome://rewards"},
+};
+
+const std::map<std::string, std::string> origins = {
+  {kBinanceScheme, "https://accounts.binance.com"},
+  {kGeminiScheme, "https://exchange.gemini.com"},
+};
+
+const std::vector<std::string> rewards_origins = {
+  "https://uphold.com",
+  "https://sandbox.uphold.com"
+};
+
+const std::string GetAuthToken(const GURL& url) {
+  std::map<std::string, std::string> parts;
+
+  for (net::QueryIterator it(url); !it.IsAtEnd(); it.Advance()) {
+    parts[it.GetKey()] = it.GetUnescapedValue();
+  }
+
+  if (parts.find("code") != parts.end()) {
+    return parts["code"];
+  } else {
+    return std::string();
+  }
+}
+
+GURL GetRedirectURL(const GURL& url,
+                    bool preserve_path = false) {
+  if (!url.is_valid()) {
+    return GURL();
+  }
+
+  const std::string scheme = url.scheme();
+
+  if (!preserve_path) {
+    return GURL(redirects.at(scheme));
+  }
+
+  std::string query;
+  std::string path = url.path();
+
+  if (url.has_query()) {
+    query = base::StrCat({
+      "?",
+      net::EscapeExternalHandlerValue(url.query())
+    });
+  }
+
+  base::ReplaceFirstSubstringAfterOffset(&path, 0, "/", "");
+
+  return GURL(
+      base::StrCat({
+        redirects.at(scheme),
+        path,
+        query
+      }));
+}
+
+void SetAuthToken(const std::string& auth_token,
+                  const std::string& oauth_scheme,
+                  content::WebContents* web_contents) {
+  if (auth_token.empty()) {
+    return;
+  }
+
+  if (oauth_scheme == kBinanceScheme) {
+#if BUILDFLAG(BINANCE_ENABLED)
+    Profile* profile =
+        Profile::FromBrowserContext(
+          web_contents->GetBrowserContext());
+    BinanceServiceFactory::GetInstance()
+      ->GetForProfile(profile)
+      ->SetAuthToken(auth_token);
+#endif
+    } else if (oauth_scheme == kGeminiScheme) {
+#if BUILDFLAG(GEMINI_ENABLED)
+    Profile* profile =
+        Profile::FromBrowserContext(
+          web_contents->GetBrowserContext());
+    GeminiServiceFactory::GetInstance()
+      ->GetForProfile(profile)
+      ->SetAuthToken(auth_token);
+#endif
+  }
+}
+
+bool IsAllowedOrigin(content::WebContents* web_contents,
+                     const base::Optional<url::Origin>& initiating_origin,
+                     const std::string& scheme) {
+  if (scheme == kRewardsScheme) {
+    bool is_allowed = std::find(
+      rewards_origins.begin(),
+      rewards_origins.end(),
+      scheme) != rewards_origins.end();
+    return is_allowed;
+  }
+
+  GURL allowed_origin(origins.at(scheme));
+  bool unallowed_origin =
+    !initiating_origin.has_value() ||
+    initiating_origin.value().GetURL() != allowed_origin ||
+    web_contents->GetLastCommittedURL().GetOrigin() != allowed_origin;
```

Comment:
Can we add these extra checks for rewards too? I think it's an oversight that they weren't there for rewards.

Reply:
yep, can do

Reply:
Same checks are now done against rewards origin

Reply:
thanks